pkg/gofr/datasource/opentsdb/opentsdb.go

```diff
@@ -0,0 +1,353 @@
+	// via one calling of /api/put method.
+	// This value is optional, and if it is not set, client.DefaultMaxPutPointsNum
+	// will be used in the opentsdb client.
+	MaxPutPointsNum int
+
+	// The detect delta number of datapoints which will be used in client.Put()
+	// to split a large group of datapoints into small batches.
+	// This value is optional, and if it is not set, client.DefaultDetectDeltaNum
+	// will be used in the opentsdb client.
+	DetectDeltaNum int
+
+	// The maximal body content length per /api/put method to insert datapoints
+	// into opentsdb.
+	// This value is optional, and if it is not set, client.DefaultMaxPutPointsNum
+	// will be used in the opentsdb client.
+	MaxContentLength int
+}
+
+// New initializes a new instance of Opentsdb with provided configuration.
+func New(config *OpenTSDBConfig) *OpentsdbClient {
+	return &OpentsdbClient{opentsdbCfg: *config}
+}
+
+func (c *OpentsdbClient) UseLogger(logger interface{}) {
+	if l, ok := logger.(Logger); ok {
+		c.logger = l
+	}
+}
+
+func (c *OpentsdbClient) UseMetrics(metrics interface{}) {
+	if m, ok := metrics.(Metrics); ok {
+		c.metrics = m
+	}
+}
+
+func (c *OpentsdbClient) UseTracer(tracer any) {
+	if tracer, ok := tracer.(trace.Tracer); ok {
+		c.tracer = tracer
+	}
+}
+
+// DefaultTransport defines the default HTTP transport settings,
+// including connection timeouts and idle connections.
+var DefaultTransport = &http.Transport{
+	MaxIdleConnsPerHost: 10,
+	DialContext: (&net.Dialer{
+		Timeout:   DefaultDialTime,
+		KeepAlive: ConnectionTimeout,
+	}).DialContext,
+}
+
+// Connect initializes an HTTP client for OpenTSDB using the provided configuration.
+// If the configuration is invalid or the endpoint is unreachable, an error is logged.
+func (c *OpentsdbClient) Connect() {
+	if c.ctx == nil {
+		c.ctx = context.Background()
+	}
+
+	span := c.addTrace(c.ctx, "Connect")
+
+	status := StatusFailed
+
+	var message string
+
+	defer sendOperationStats(c.logger, time.Now(), "Connect", &status, &message, span)
+
+	c.opentsdbCfg.OpentsdbHost = strings.TrimSpace(c.opentsdbCfg.OpentsdbHost)
+	if c.opentsdbCfg.OpentsdbHost == "" {
+		c.logger.Errorf("the OpentsdbEndpoint in the given configuration cannot be empty.")
+	}
+
+	// Use custom transport settings if provided, otherwise, use the default transport.
+	transport := c.opentsdbCfg.Transport
+	if transport == nil {
+		transport = DefaultTransport
+	}
+
+	c.client = &http.Client{
+		Transport: transport,
+	}
+
+	// Set default values for optional configuration fields.
+	if c.opentsdbCfg.MaxPutPointsNum <= 0 {
+		c.opentsdbCfg.MaxPutPointsNum = DefaultMaxPutPointsNum
+	}
+
+	if c.opentsdbCfg.DetectDeltaNum <= 0 {
+		c.opentsdbCfg.DetectDeltaNum = DefaultDetectDeltaNum
+	}
+
+	if c.opentsdbCfg.MaxContentLength <= 0 {
+		c.opentsdbCfg.MaxContentLength = DefaultMaxContentLength
+	}
+
+	// Initialize the OpenTSDB client with the given configuration.
+	c.tsdbEndpoint = fmt.Sprintf("http://%s", c.opentsdbCfg.OpentsdbHost)
```

Comment:
According to the OpenTSDB REST API documentation, it uses a basic HTTP-based API. All available examples also reference HTTP rather than HTTPS, which suggests that HTTPS is not currently supported. This is likely because OpenTSDB was originally designed for internal use, where the need for encrypted communication is less critical.